File: service/src/main/java/org/wcl/mfn/service/tools/ContractCalculatorService.java
package org.wcl.mfn.service.tools;

import org.springframework.stereotype.Service;
import org.wcl.mfn.entities.contract.calculator.*;
import org.wcl.mfn.exceptions.validation.InvalidParameterException;

import java.util.List;
import java.util.stream.IntStream;


@Service
public final  class ContractCalculatorService {
    private static final double BONUS_PERCENTAGE_OF_SALARY = 0.75;
    private static final int MIN_YEARS = 2;
    private static final int MAX_YEARS = 6;
    public static final double SALARY_SCALING_FACTOR = 1.05;

    public List<SuggestedContract> suggestedContracts(int requestedSalary, int requestedBonus) throws InvalidParameterException {
        if (requestedSalary <= 0) {
            throw new InvalidParameterException("Invalid requested salary - cannot be negative or zero");
        }
        if (requestedBonus <= 0) {
            throw new InvalidParameterException("Invalid requested bonus - cannot be negative or zero");
        }

        return IntStream.rangeClosed(MIN_YEARS,MAX_YEARS).mapToObj(n -> generateSuggestedContract(n, requestedSalary, requestedBonus)).toList();
    }

    private SuggestedContract generateSuggestedContract(int numberOfYears, int requestedSalary, int requestedBonus) {
        final var bonusPerYear = (int)Math.ceil((requestedBonus / (double)numberOfYears));
        final var startingSalary = generateStartingSalary(requestedSalary, requestedBonus, bonusPerYear);
        final var yearlyFigures = generateYearlyFigures(numberOfYears, startingSalary, bonusPerYear);

        return new SuggestedContract(
                numberOfYears,
                yearlyFigures.stream().mapToInt(SuggestedYearlyContract::salary).sum(),
                bonusPerYear,
                yearlyFigures.stream().mapToInt(SuggestedYearlyContract::yearlyRemuneration).sum(),
                yearlyFigures
        );
    }

    private int generateStartingSalary(int requestedSalary, int requestedBonus, int bonusPerYear) {
        if (requestedSalary > requestedBonus) {
            return requestedSalary;
        } else {
            return (int)Math.ceil(bonusPerYear * BONUS_PERCENTAGE_OF_SALARY);
        }
    }

    private List<SuggestedYearlyContract> generateYearlyFigures(int numberOfYears, int startingSalary, int bonusPerYear) {
        return IntStream.rangeClosed(1, numberOfYears)
                .mapToObj(n -> generateYearlyFigure(n, startingSalary, bonusPerYear)).toList();
    }

    private SuggestedYearlyContract generateYearlyFigure(int year, int startingSalary, int bonusPerYear)  {
        int salary = startingSalary;

        if (year > 1) {
            salary = (int)Math.ceil(startingSalary * (Math.pow(SALARY_SCALING_FACTOR,year - 1)));
        }
        return new SuggestedYearlyContract(year, salary, salary + bonusPerYear);
    }
}

File: service/src/main/kotlin/org/wcl/mfn/service/tools/ContractCalculatorService.kt
package org.wcl.mfn.service.tools

import org.springframework.stereotype.Service
import org.wcl.mfn.entities.contract.calculator.*
import org.wcl.mfn.exceptions.validation.InvalidParameterException
import kotlin.math.*

@Service
class ContractCalculatorService {
    fun suggestedContracts(requestedSalary: Int, requestedBonus: Int): List<SuggestedContract> {
        if (requestedSalary <= 0) {
            throw InvalidParameterException("Invalid requested salary - cannot be negative or zero");
        }
        if (requestedBonus <= 0) {
            throw InvalidParameterException("Invalid requested bonus - cannot be negative or zero");
        }

        return generateSuggestedContracts(requestedSalary, requestedBonus)
    }

    private fun generateSuggestedContracts(requestedSalary: Int, requestedBonus: Int): List<SuggestedContract> {
        val yearsList = IntRange(2, 6).toList()
        return yearsList.stream().map { n -> generateSuggestedContract(n, requestedSalary, requestedBonus) }.toList()
    }

    private fun generateSuggestedContract(numberOfYears: Int, requestedSalary: Int, requestedBonus: Int): SuggestedContract {
        val bonusPerYear = ceil((requestedBonus / numberOfYears.toDouble())).toInt()
        val startingSalary = generateStartingSalary(requestedSalary, requestedBonus, bonusPerYear)
        val yearlyFigures = generateYearlyFigures(numberOfYears, startingSalary, bonusPerYear)

        return SuggestedContract(
            numberOfYears,
            yearlyFigures.stream().map { yf -> yf.salary }.toList().sum(),
            bonusPerYear,
            yearlyFigures.stream().map { yf -> yf.yearlyRemuneration }.toList().sum(),
            yearlyFigures
        )
    }

    private fun generateStartingSalary(requestedSalary: Int, requestedBonus: Int, bonusPerYear: Int): Int {
        return if (requestedSalary > requestedBonus) {
            requestedSalary
        } else {
            ceil(bonusPerYear * 0.75).toInt()
        }
    }

    private fun generateYearlyFigures(numberOfYears: Int, startingSalary: Int, bonusPerYear: Int): List<YearlyFigure> {
        return IntRange(1, numberOfYears).toList().stream()
            .map { n -> generateYearlyFigure(n, startingSalary, bonusPerYear) }.toList()
    }

    private fun generateYearlyFigure(year: Int, startingSalary: Int, bonusPerYear: Int): YearlyFigure {
        val salary: Int = if (year == 1) {
            startingSalary
        } else {
            ceil(startingSalary.toDouble() * (1.05.pow(year - 1))).toInt()
        }
        return YearlyFigure(year, salary, salary + bonusPerYear)
    }
}
